src/sort.rs

```diff
@@ -0,0 +1,264 @@
+    /// tree.root_mut().sort_by(|a, b| b.cmp(a));
+    /// assert_eq!(
+    ///     vec![&'d', &'c', &'b'],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by<F>(&mut self, mut compare: F)
+    where
+        F: FnMut(&T, &T) -> Ordering,
+    {
+        if self.has_children() {
+            let (unsorted, sorted) = self.sort_handler(|nodes| {
+                nodes.sort_by(|(_, a), (_, b)| compare(a, b));
+            });
+
+            self.swap(unsorted, sorted);
+        }
+    }
+
+    /// Sort children by value's key in ascending order using a key extraction function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ego_tree::tree;
+    ///
+    /// let mut tree = tree!("1a" => { "2b", "4c", "3d" });
+    /// tree.root_mut().sort_by_key(|a| a.split_at(1).0.parse::<i32>().unwrap());
+    /// assert_eq!(
+    ///     vec!["2b", "3d", "4c"],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| *n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by_key<K, F>(&mut self, mut f: F)
+    where
+        F: FnMut(&T) -> K,
+        K: Ord,
+    {
+        if self.has_children() {
+            let (unsorted, sorted) = self.sort_handler(|nodes| {
+                nodes.sort_by_key(|(_, value)| f(value));
+            });
+
+            self.swap(unsorted, sorted);
+        }
+    }
+
+    /// Sort children by their NodeId in ascending order. The purpose is to restore the original order.
+    ///
+    /// This method is a shorthand for calling `sort_by_id` with the `Ord::cmp` method.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ego_tree::tree;
+    ///
+    /// let mut tree = tree!('a' => { 'd', 'c', 'b' });
+    /// tree.root_mut().sort();
+    /// assert_ne!(
+    ///     vec![&'d', &'c', &'b'],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// tree.root_mut().sort_id();
+    /// assert_eq!(
+    ///     vec![&'d', &'c', &'b'],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_id(&mut self) {
+        self.sort_by_id(|a, b| a.cmp(&b));
+    }
+
+    /// Sort children by their NodeId's index using a comparison function.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ego_tree::tree;
+    ///
+    /// let mut tree = tree!('a' => { 'd', 'b', 'c' });
+    /// tree.root_mut().sort_by_id(|a, b| b.cmp(&a));
+    /// assert_eq!(
+    ///     vec![&'c', &'b', &'d'],
+    ///     tree.root()
+    ///         .children()
+    ///         .map(|n| n.value())
+    ///         .collect::<Vec<_>>(),
+    /// );
+    /// ```
+    pub fn sort_by_id<F>(&mut self, mut compare: F)

```

Comment:
To limit the API surface, I would also suggest only three methods:
* `sort_by` which uses a closure taking two `NodeRef` and can be used to implement anything else,
* `sort_by_key` using a closure taking a single `NodeRef` and producing a key and
* `sort` which sorts by value.

Things like `sort_id_by` are then trivially implemented by calling `node.id()` in `sort_by_key`.

Reply:
Yes, a sensible addition even without this API.